### PR TITLE
feat: Add WIP state checkpointing before compaction

### LINKED ISSUE #160: RFC: Session Continuity - Checkpoint WIP State Before Compaction
## Summary

When Claude Code sessions run out of context and compact, the continuation summary loses work-in-progress state (uncommitted edits, current file/line position, error being debugged). This forces continuation sessions to re-discover context that was already known, wasting tokens and causing friction.

## Problem / Motivation

Session compaction summaries capture *what was done* but not *what was in progress*:

- **Uncommitted edits**: If mid-way through editing a file, the specific changes and their purpose are lost
- **Current position**: Which file/function/line was being worked on
- **Error state**: What error was being debugged, what was already tried
- **Todo progress**: Which item in a multi-step task was active

**Evidence**: Workflow analysis identified 5 continuation sessions in 7 days that needed to re-discover context that was partially completed. Example: A session continued with "First task: Fix proptest_tests.rs import error" but didn't know the user was already mid-way through updating test expectations in `request_tests.rs`.

## Context

- **Discovered during**: Workflow analysis after completing PR #293 in rust-genai
- **Relevant files**: `~/.claude/commands/work.md`, potentially new `~/.claude/hooks/`
- **Related issues/PRs**: None directly, this is a new observation

## Proposed Solution

Enhance the `/work` workflow to checkpoint WIP state that survives compaction:

### Option A: WIP State File
Create `.claude/wip-state.json` in the project directory:
```json
{
  "work_id": "issue-272",
  "current_todo": "Fix dual wire format issue",
  "current_todo_status": "in_progress",
  "files_modified": ["agent_config.rs", "request_tests.rs"],
  "last_error": "test_thinking_summaries_serialization failed - expected lowercase",
  "last_action": "Edit request_tests.rs line 111",
  "checkpoint_time": "2026-01-04T20:00:00Z"
}
```

The `/work --attach` command would read this file to restore context.

### Option B: Git Stash with Message
Use `git stash push -m "WIP: [work:issue-272] Fixing dual wire format"` to preserve uncommitted changes with context in the stash message.

### Option C: Hook-Based Checkpointing
Add a pre-compaction hook (if Claude Code supports it) that writes state before context is lost.

## Assumptions

| Assumption | Confidence | Impact if Wrong |
|------------|------------|-----------------|
| Continuation sessions can read project files | High | Would need alternative storage |
| WIP state is useful even without uncommitted code | Medium | May only help partial cases |
| `/work` workflow is the right place for this | Medium | Could be a standalone hook instead |
| Claude Code doesn't already have a better mechanism | Low | This RFC may be unnecessary |

## Open Questions

1. **Does Claude Code have hooks for pre-compaction?** If so, Option C is cleanest.
2. **Should WIP state include actual file diffs?** Would be more complete but larger.
3. **How to handle multiple worktrees?** Each would need its own WIP state.

## Actionable Requirements

| # | Requirement | Owner | Blocked By |
|---|-------------|-------|------------|
| 1 | Research Claude Code compaction hooks | Claude | - |
| 2 | Define WIP state schema | Claude | #1 |
| 3 | Implement checkpoint writing in /work | Claude | #2 |
| 4 | Implement checkpoint reading in /work --attach | Claude | #3 |
| 5 | Add cleanup on /work completion | Claude | #4 |

## Test Requirements

- **Manual**: Start work, let session compact, continue with `/work --attach`, verify context restored
- **Edge cases**: 
  - Multiple todos in progress
  - Worktree-based parallel work
  - Work abandoned (no cleanup)

## Implementation Checklist

- [ ] Research Claude Code compaction behavior and hooks
- [ ] Design WIP state schema
- [ ] Update `/work` to write checkpoint on todo state changes
- [ ] Update `/work --attach` to read and display WIP context
- [ ] Add cleanup when work completes (merge/close)
- [ ] Document in CLAUDE.md or workflow docs

### PR DESCRIPTION
**Pull request by evansenter**
## Summary

Implements session continuity by preserving work-in-progress state across context compactions. This addresses the problem where Claude Code sessions lose uncommitted edits, current position, error state, and todo progress when context is summarized.

**Architecture:**
- **PreCompact hook** → Extracts git state (branch, modified files, work ID) → Publishes to event bus
- **SessionStart hook** → Detects `source == "compact"` → Retrieves and displays WIP checkpoint
- **/work command** → Instructs Claude to checkpoint progress periodically

**Changes:**
- Add `pre-compact.sh` hook that captures and publishes WIP state
- Update `session-start.sh` to restore WIP checkpoint on compact resume
- Register PreCompact hook in `settings.json`
- Extend `/work` with checkpointing instructions and cleanup on completion
- Add comprehensive tests (6 new tests for pre-compact.sh)

## Test plan

- [x] Run `make check` - all 37 tests pass
- [ ] Trigger manual compaction (if supported) and verify checkpoint published
- [ ] Resume after compaction and verify WIP state displayed
- [ ] Verify graceful degradation when event-bus-cli unavailable

Closes #160

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: home/.claude/commands/work.md [hunk 1]
```diff
@@ -115,6 +117,26 @@ Mark first task `in_progress`.
...
+
+**When to checkpoint:** After completing any significant todo, or before long-running operations.
+
+**How to checkpoint:**
+```
+mcp__event-bus__publish_event(
+  event_type: "wip_progress",
+  payload: "[work:<ID>] Completed: <task>. In progress: <next_task>. Remaining: <count>",
+  session_id: "<your-session-id>",
+  channel: "session:<your-session-id>"
```
Comment (evansenter) [inline home/.claude/commands/work.md]: I don’t think this is the right format, I think channel should be session:<your-display-name>. Check for me.

File: home/.claude/hooks/session-start.sh [hunk 2]
```diff
@@ -69,3 +70,23 @@ if [[ -n "$EVENTS" && "$EVENTS" != "No events" && "$EVENTS" != "No new events" ]
...
     echo "</recent-events>"
 fi
+
+# If resuming after compaction, fetch and display WIP checkpoint
+if [[ "$SOURCE" == "compact" ]]; then
+    # Fetch the most recent wip_checkpoint event for this session
+    # Note: CLI only has --exclude-types, so we filter with grep post-fetch
+    WIP_EVENT=$(event-bus-cli events \
+        --session-id "$SESSION_ID" \
+        --channel "session:${SESSION_ID}" \
```
Comment (evansenter) [inline home/.claude/hooks/session-start.sh]: i don’t think this is right
